modules/navigation/navigation_path_query_result_3d.cpp

```diff
@@ -0,0 +1,54 @@
+/*************************************************************************/
+/*  navigation_path_query_result_3d.cpp                                  */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "navigation_path_query_result_3d.h"
+
+void NavigationPathQueryResult3D::set_path(const Vector<Vector3> &p_path) {
+	path = p_path;
+}
+
+const Vector<Vector3> &NavigationPathQueryResult3D::get_path() const {
+	return path;
+}
+
+void NavigationPathQueryResult3D::set_path_length(real_t p_path_length) {
+	path_length = p_path_length;
+}
+
+real_t NavigationPathQueryResult3D::get_path_length() const {
+	return path_length;
+}
+
+void NavigationPathQueryResult3D::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("get_path"), &NavigationPathQueryResult3D::get_path);
+	ClassDB::bind_method(D_METHOD("get_path_length"), &NavigationPathQueryResult3D::get_path_length);
+
+	ADD_SIGNAL(MethodInfo("result_changed"));
```

Comment:
It seems a little odd to me to register for a signal on a temporary result object.  How would this look in GDScript?

Ideally, I'd like to do something like:

``` gdscript

var query := NavigationPathQueryParameters3D();
# query.target_position = ...
# fill in parameters...

var result := await NavigationServer3D.query_path(query);

```

Do we know if GDScript has the concept of a `Promise` or `Awaitable` instead of only waiting for a signal?

Reply:
The result is no longer a throw-away object. The idea is to create both objects and reuse them. The signal is basically there so users can connect for updates in preperation for async path queries. Since this is not added now it is not so useful cause as soon as a user uses the query_path() function the result is updated.

Reply:
If these objects are intended to stick around, maybe we should just have one of them to avoid confusion with the physics APIs (which look similar but don't work the same way).  It also simplifies the API by not requiring the user to keep two objects around and remember to always pair them with each other.

So, we'd have `NavigationPathQuery3D` which contains the properties of both `NavigationPathQueryParameters3D` and `NavigationPathQueryResult3D`.

Then you'd call it like:

``` gdscript
var query := NavigationPathQuery3D();

# fill in parameters...
query.target_location = Vector3(...);
query.current_location = self.global_position;

# run the query
NavigationServer3D.query_path(query);

# wait for the query to complete
await query.completed;

# get back the final path
var path = query.get_path();
```

Reply:
Physics has no result object reuse and it is actually quite wasteful e.g. new object with each query or using dictionary.

The split makes sense cause you do not always want both objects created at an equal count. E.g. I have an actor manager that reuses the same parameter object by just updating locations in a timesliced loop or I have multiple parameters for gameplay situations on my actor but it only uses one result for the pathfinding.

Reply:
Agree with @DarkKilauea on this being odd to have the signal as part of the query. Could this be a signal on the `NavigationAgent` like `velocity_computed`?
Given that this is now adapted to work in an async framework, can the signal return the result (separate from the query)?

I don't know if there are potentials for race conditions of several queries before the first is returned.

Reply:
The query needs to work without any NavigationAgent node involved so where else to place a signal? Can copy the way how avoidance callbacks solve this but those use no signal and solve this by registering callbacks first and then calling the function on the object directly. The NavigationAgent emits the velocity_computed after the internal avoidance callback functions gets called.

Reply:
This sounds fine imo, I would just copy the technique from avoidance callbacks. :+1: